lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb

```diff
@@ -1,25 +1,23 @@
 # frozen_string_literal: true
 
-require 'active_record/connection_adapters/abstract/transaction'
-
 module ActiveRecord
   module ConnectionAdapters
-
-    # NOTE(joey): This is a very sad monkey patch. Unfortunately, it is
-    # required in order to prevent doing more than 2 nested transactions
-    # while still allowing a single nested transaction. This is because
-    # CockroachDB only supports a single savepoint at the beginning of a
-    # transaction. Allowing this works for the common case of testing.
     module CockroachDB
       module TransactionManagerMonkeyPatch
-        def begin_transaction(options={})
-          @connection.lock.synchronize do
-            # If the transaction nesting is already 2 deep, raise an error.
-            if @connection.adapter_name == "CockroachDB" && @stack.is_a?(ActiveRecord::ConnectionAdapters::SavepointTransaction)
-              raise(ArgumentError, "cannot nest more than 1 transaction at a time. this is a CockroachDB limitation")
-            end
-          end
-          super(options)
+        # Capture ActiveRecord::SerializationFailure errors caused by
+        # transactions that fail due to serialization errors. Failed
+        # transactions will be retried until they pass or the max retry limit is
+        # exceeded.
+        def within_new_transaction(options = {})
+          attempts = options.fetch(:attempts, 0)
+          super
+        rescue ActiveRecord::SerializationFailure => error
+          raise if attempts >= @connection.max_transaction_retries
+
+          attempts += 1
+          sleep_seconds = (((2 ** attempts) * 100) + rand( 100 - 1 ) + 1) / 1000.0
```

Comment:
Why is this random?

Reply:
It's based on the example from the docs: https://www.cockroachlabs.com/docs/v19.2/transactions.html#client-side-intervention-example.

Reply:
I guess the idea is to desync if there are two calls that start at the same time. Clean up the math a bit and it's good.

Reply:
What should we change the math to? It's currently the same math that's used in the example.

```python
# This is a retry error, so we roll back the current transaction
# and sleep for a bit before retrying. The sleep time increases
# for each failed transaction.  Adapted from
# https://colintemple.com/2017/03/java-exponential-backoff/
conn.exec('ROLLBACK');
sleep_ms = int(((2**n) * 100) + rand( 100 - 1 ) + 1)
```

Reply:
I'd prefer `(2 ** attempts + rand()) / 10`. It's also not clear to me why the +1 is there.

Reply:
Yep, that's a lot cleaner and gets us the same results. Updated in 64d450f5007f59a55f1431be0532ca4d294c75c9.